### ISSUE TITLE
[$250] Send invoice - System message for invoice edit is missing "Reply in thread" for invoice sender

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#43797]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.84-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y 
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Go to staging.new.expensify.com
2. Go to FAB > Send invoice
3. Send invoice to User B
4. Go to transaction thread
5. Edit Amount
6. [User A and User B] Right click on the system message
7. Note that "Reply in thread" is missing for invoice sender but present for invoice receiver
8. [User B] Send a reply to the system message
9. [User A] Open the reply in thread
10. {User A] Click on the thread header subtitle
11. {User A] Click on the thread header subtitle again

## Expected Result:
In Step 7, "Reply in thread" should be available for both invoice sender and receiver
In Step 10, the invoice thread should have no visual issue
In Step 11, user should be navigated to the main chat

## Actual Result:
In Step 7, "Reply in thread" is missing for invoice sender but present for invoice receiver
In Step 10, the invoice details appear broken above existing invoice details
In Step 11, user is navigated to the same invoice thread instead of the main chat

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/650bc6bb-7e6a-401d-91ae-12d5bfbb759c

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0111ac375ebfe2baa1</li>
        <li>Upwork Job ID: 1803175554822916116</li>
        <li>Last Price Increase: 2024-08-08</li>
<li>Automatic offers: </li>
<ul>
<li>rayane-djouah | Reviewer | 103213577</li>
<li>truph01 | Contributor | 103213579</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @dylanexpensify</details>

**Comment by lanitochka17** [comment]
@dylanexpensify  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by truph01** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
In Step 7, "Reply in thread" is missing for invoice sender but present for invoice receiver
In Step 10, the invoice details appear broken above existing invoice details
In Step 11, user is navigated to the same invoice thread instead of the main chat
### What is the root cause of that problem?
When we create a send invoice, we don't pass the ID of `iouAction` as a parameter in the `SendInvoice` API. It leads after the API is complete, the iou report has two `IOU` action

https://github.com/Expensify/App/blob/1e0831eabdadc9380b7bc1d8a194ae05102c9736/src/libs/actions/IOU.ts#L3545

And then `getOriginalReportID` returns the wrong `reportID` for the system message because we get all reportActions of the iou report in Onyx which makes `getOneTransactionThreadReportID ` function returns `undefined`.

https://github.com/Expensify/App/blob/1e0831eabdadc9380b7bc1d8a194ae05102c9736/src/libs/ReportUtils.ts#L6063

It doesn't happen in the `ReportScreen` because when we get the report actions here, it doesn't include the optimistic iou action

https://github.com/Expensify/App/blob/1e0831eabdadc9380b7bc1d8a194ae05102c9736/src/pages/home/ReportScreen.tsx#L261
### What changes do you think we should make in order to solve the problem?
We should pass the `reportActionID` of `iouAction` as a parameter of the `SendInvoice` API and BE will use this `reportActionID` to create the `iouAction` and return data for the user.

https://github.com/Expensify/App/blob/1e0831eabdadc9380b7bc1d8a194ae05102c9736/src/libs/actions/IOU.ts#L3545

To do that we should
1. In `getSendInvoiceInformation` function, we will return `iouReportActionID: iouAction.reportID` here
https://github.com/Expensify/App/blob/1e0831eabdadc9380b7bc1d8a194ae05102c9736/src/libs/actions/IOU.ts#L1844
### What alternative solutions did you explore? (Optional)

2. Get the `iouReportActionID` when we call `getSendInvoiceInformation` function here
https://github.com/Expensify/App/blob/1e0831eabdadc9380b7bc1d8a194ae05102c9736/src/libs/actions/IOU.ts#L3542

and then pass it as a parameter of `SendInvoice` API here
https://github.com/Expensify/App/blob/1e0831eabdadc9380b7bc1d8a194ae05102c9736/src/libs/actions/IOU.ts#L3545

3. We need a BE change here that will use `iouReportActionID` param to generate the the iouAction.

OPTIONAL: The created action of the transaction thread report is also duplicated. The RCA is the same with the `iouAction`, we can add the same fix for this action as well.

**Comment by tsa321** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Send invoice - System message for invoice edit is missing "Reply in thread" for invoice sender and some visual issues related to transaction details.

### What is the root cause of that problem?

When user send an invoice to the server returns new IOU report action which have different report action than optimistic report action. When user right click the system message (which is message from transactions thread and not from report which have many reportpreview list report. If the report only has exactly 1 report preview it will immediately displays transaction details. So the system message when user edit IOU is actually from child report but displayed at parent because the reportpreview is only 1).

Since server return different IOU, the report actions onyx will have more than 1 IOU report actions (optimistic and the report action returned by server), it will fails to give correct result in:

https://github.com/Expensify/App/blob/b3f688b8213a95bdaf4a97d1fa389f396c8be25d/src/libs/ReportActionsUtils.ts#L915-L917

`getOneTransactionThreadReportID` will early return in the code above and gives undefined transaction thread.

and will eventually fails to get correct report action in here:

https://github.com/Expensify/App/blob/b3f688b8213a95bdaf4a97d1fa389f396c8be25d/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L131-L136

and `replyInThread` context menu will eventually hidden because shouldShow of action menu give incorrect result.  

### What changes do you think we should make in order to solve the problem?

We could remove onyx optimistic IOU report actions because server returns new/different IOU report actions, in here:

https://github.com/Expensify/App/blob/b3f688b8213a95bdaf4a97d1fa389f396c8be25d/src/libs/actions/IOU.ts#L1035-L1041

inside successData in `buildOnyxDataForInvoice`, 

We set the 
```
[transactionThreadCreatedReportAction.reportActionID]: null
....

[iouCreatedAction.reportActionID]: null
[iouAction.reportActionID]: null,
```
We set it to null, to remove the optimistic iou report actions...

**Comment by dylanexpensify** [comment]
added to project

**Comment by tsa321** [comment]
@rayane-djouah 

> Why is the "Reply in thread" option disabled for invoice sender?

Basically because there are two IOU report action after user send invoice (one from optimistic action, one from server, this two IOU have different report action ids). 
The optimistic IOU is filtered by this function, so  it is not displayed in report actions list:

https://github.com/Expensify/App/blob/f618a94c486b80aebab584c52288a7013ffad64a/src/pages/home/ReportScreen.tsx#L261

When report actions in view/list only have 1 IOU report actions it will display transaction details (this is already correct).
If it is more than one it will displayed as REPORTPREVIEW for each IOU report actions.

Also, when only 1 IOU report actions is displayed, the child transactions details thread reportactions will be merged into parent report report actions(which is the report which don't has "Reply in thread").
So basically the report actions of  report which doesn't have a "Reply in thread"  consist of report actions from that report and child transaction thread report. 

Because of the getContinuousReportActionChain which is success to filter the optimistic IOU report action, but when user  right click to show report acitons context menu item it fails to display `Reply in Thread`.  This is because:

https://github.com/Expensify/App/blob/b3f688b8213a95bdaf4a97d1fa389f396c8be25d/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L205-L207

The `getOriginalReportID` fails to get the correct child (original) report id / the transaction thread report which is the report that show the system message about money request update /  the report which is the report actions are merged into parent if parent only has 1 IOU report action. 

This is because:

https://github.com/Expensify/App/blob/f618a94c486b80aebab584c52288a7013ffad64a/src/libs/ReportUtils.ts#L6066-L6070

`getOneTransactionThreadReportID` :

https://github.com/Expensify/App/blob/f618a94c486b80aebab584c52288a7013ffad64a/src/libs/ReportActionsUtils.ts#L920-L922

Fails to returns correct childreportid / trasaction thread report id, because early return in that if clause. 
This is because there are two IOU report actions (one from optimistic update and one from backend, these two report actions have different reportActionID).

Because of incorrect `originalReportID`, the reportActions here:

https://github.com/Expensify/App/blob/f618a94c486b80aebab584c52288a7013ffad64a/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L131-L144

the reportAction here will be undefined, and `ContextMenuActions.filter` will filter `Reply In Thread` context menu because the context menu give false result in `contextAction.shouldShow`.

I have updated my [proposal](https://github.com/Expensify/App/issues/43797#issuecomment-2173688182) in RCA section.


> Why is there a duplicate report action for the Why is there a duplicate report action for the invoice details??

Because after user send invoice the report action of type `CREATE` and `IOU` from optimistic update and from back end are preserved/stored. Both have different report action ids. The details is representation of CREATE report action, when there is two create report action, there is a glitch when showing two invoice details.

> Why is the user navigated to the same invoice thread instead of the main chat?

because the system message of IOU request upate/edit are from child thread transaction report. The system message is displayed/merged into parent report because the parent has exactly one IOU report action (because report screen success filtering the optimistic IOU report action).
basically the system message is from child transaction thread report.

**Comment by truph01** [comment]
@rayane-djouah 

> Why is the "Reply in thread" option disabled for invoice sender?

This because the originalReportID is wrong as I mentioned in my proposal and then we can't get the current reportAction in `BaseReportActionContextMenu`

> Why is there a duplicate report action for the invoice details?

Because BE generate an iou action with a random `reportActionID` instead of using the `reportActionID` from front-end, after `SendInvoice` API is complete we have two iou action, one from server, one is optimistic iou action.

> Why is the user navigated to the same invoice thread instead of the main chat?

This is not the same chat. The first chat is the transaction thread report that is the current parent report of the system message thread. The second is the invoice room but this report has only one transaction and then this report display as combine report that will display all report actions of invoice room and all report actions of the transaction thread. So this case isn't a bug.

**Comment by dylanexpensify** [comment]
@cristipaval to review and confirm

**Comment by cristipaval** [comment]
I think @truph01 can work on new issues once they merge all the ongoing ones. @dylanexpensify, can you confirm? 🙏

**Comment by cristipaval** [comment]
Thanks for pushing this one, @rayane-djouah! 🙏 

I think we better do the backend changes first. The thing is that this is not a high priority issue at I won't be able to work on it this week.

**Comment by cristipaval** [comment]
I might be able to work on the backend changes for this one at the end of the week.

**Comment by dylanexpensify** [comment]
thanks @cristipaval!

**Comment by tsa321** [comment]
@rayane-djouah If we want to handle the similar errors in Front End based on [this message](https://expensify.slack.com/archives/C01GTK53T8Q/p1720533672162139?thread_ts=1720533146.376329&cid=C01GTK53T8Q), we could remove the optimistic created report actions (and optimistic IOU report action) after API request finished.

**Comment by cristipaval** [comment]
Not overdue, Melv.

**Comment by cristipaval** [comment]
The backend PRs are in review

**Comment by cristipaval** [comment]
Both backend PRs are merged, they just need to hit staging.

**Comment by cristipaval** [comment]
Assigned @truph01 in the meantime. Just keep in mind that you can fully test when the backend PRs are in staging. I'll post an update here when they are

**Comment by truph01** [comment]
Daily updates: We are still waiting for the backend PR to hit staging.

**Comment by cristipaval** [comment]
> Daily updates: We are still waiting for the backend PR to hit staging.

They are in staging. Feel free to test and let me know if it works as expected.

**Comment by truph01** [comment]
@cristipaval I try to send the param named ```iouReportActionID``` to BE when calling API ```SendInvoice``` but there is no proper action returned by BE. Can you help check?

**Comment by cristipaval** [comment]
@truph01, are you saying that the IOU action is not added in the `onyxData` of the `SendInvoice` command response? I'm trying to understand what you mean by
> but there is no proper action returned by BE

**Comment by truph01** [comment]
>  are you saying that the IOU action is not added in the onyxData of the SendInvoice command response? I'm trying to understand what you mean by

Yes

**Comment by truph01** [comment]
Also, I found out another bug: The OpenReport API return 2 created action in invoice room

**Comment by cristipaval** [comment]
> @cristipaval - Could you please provide us with the exact names of the parameters that you added to the endpoint?

<img width="831" alt="Screenshot 2024-07-29 at 15 16 14" src="https://github.com/user-attachments/assets/068812c0-30cf-41c2-bcdf-24db3e325368">

**Comment by truph01** [comment]
> Also, I found out another bug: The OpenReport API return 2 created action in invoice room

Have you checked this bug? 

FYI, it leads to the behavior:

![image](https://github.com/user-attachments/assets/fed1896b-e7f5-4ce3-b27d-0d0d62cc39cb)

**Comment by truph01** [comment]
Also, the step 11:

> In Step 11, user is navigated to the same invoice thread instead of the main chat

I believe this is expected behavior. We have previously closed similar issues related to the one transaction report flow. Since the system message thread is a child of the transaction thread report, clicking on the subtitle header in step 11 should navigate to the transaction thread report.

**Comment by truph01** [comment]
Bump @cristipaval about my comment [above](https://github.com/Expensify/App/issues/43797#issuecomment-2258111781)

**Comment by cristipaval** [comment]
I haven't yet, but I'm getting closer to wrapping up my other projects and then jump again into Invoicing stuff

**Comment by truph01** [comment]
@cristipaval Do you have any update on it?

**Comment by cristipaval** [comment]
I'll investigate this one later today

**Comment by cristipaval** [comment]
> > Also, I found out another bug: The OpenReport API return 2 created action in invoice room
> 
> @cristipaval Have you checked this bug?

@truph01, I think you get 2 `CREATED` report actions when you click a preview of an invoice from the invoice room, correct? It is the expected behavior. It happens the same when you open an expense report with a single expense on it. 
Because the invoice has 1 transaction on it, the user is navigated straight to the transaction thread and the `OpenReport` API command returns the report actions of both the invoice report and the transaction thread report.

**Comment by truph01** [comment]
@cristipaval I checked and saw that for exepnse report when it only has one transaction, BE still returns only one `CREATED` action. About the report actions in combine report, we will merge the actions from the money request report and transaction thread report [here](https://github.com/Expensify/App/blob/4bac1000e36d0a8ba5fba945e87053d7b8cf2ef2/src/pages/home/report/ReportActionsView.tsx#L220)

**Comment by cristipaval** [comment]
Ok, I haven't found any backend logic to return only one CREATED action when merging the actions from the expense report and transaction thread. I'll investigate again today.

**Comment by truph01** [comment]
@cristipaval Do you have any updates on this?

**Comment by cristipaval** [comment]
I investigated and I don't see code in the backend to filter out any CREATED action. I also tested with the App and verified that when you open an expense report with only one transaction, you get the CREATED action for both the expense report and the transaction thread report as well. cc @NikkiWines as I know you worked on merging the transaction thread with the expense report.

**Comment by NikkiWines** [comment]
Hmm, yeah all the filtering done for the oneTransactionReport view is done in the front-end afaik

**Comment by cristipaval** [comment]
Thanks @NikkiWines! So @truph01, let me know if you need anything else from the backend in order to move with the implementation for this issue.

**Comment by truph01** [comment]
@cristipaval 

- There are 2 CREATED actions in send invoices room. I think it is BE bug, because it should be just 1 CREATED action:

![image](https://github.com/user-attachments/assets/4a606540-61d4-4bb5-b70c-149c57bf3cfd)

- FYI, there is only 1 CREATED action in expense report:

![image](https://github.com/user-attachments/assets/83fa0fb4-d8cd-4369-bb16-c8a97fabf855)

**Comment by cristipaval** [comment]
You are talking about the invoice room, but you are showing me screenshots of the invoice report. 
Now I understand the issue. There are 2 CREATED report actions for the invoice report PLUS the one that belongs to the transaction thread. I could reproduce.

Pushing a fix asap.

For your clarity:
- **invoice room** is a chat report (report.type = chat) created between the sender workspace and the invoice receiver. Here we have report previews of the invoices sent from the sender to the receiver. It is similar to the policy chats.
- **invoice report** is the actual invoice sent to the receiver. It is an invoice report (report.type = invoice) containing a transaction. It is similar to the expense reports.
- **transaction thread report** is a chat report; its parent is the invoice report. It is exactly the same as the ones linked to the expense reports.

**Comment by truph01** [comment]
@cristipaval Thanks for your information.

**Comment by cristipaval** [comment]
Thanks for bearing with me! I'll let you know when the backend is ready

**Comment by cristipaval** [comment]
Also, could you please give me a list of the IDs of the invoice reports that you created with 2 CREATED actions? I would run a query to remove the duplicates.

**Comment by truph01** [comment]
@cristipaval The invoice report ID with 2 CREATED actions is ```7477380290328742```

**Comment by cristipaval** [comment]
Adding @madmax330 to this issue to keep him in the loop, as I might disappear on parental leave anytime soon

**Comment by cristipaval** [comment]
TLDR: 
- When [this Auth PR](https://github.com/Expensify/Auth/pull/12095) gets deployed, @truph01 is going to open an App PR to fix this issue. 
- When this issue closes, we need to also test the following issues that have the same root cause:
  - https://github.com/Expensify/App/issues/43577

**Comment by cristipaval** [comment]
We are waiting for the Auth PR to be deployed.

**Comment by cristipaval** [comment]
@truph01, the backend is ready. Just to let you know, this is fixed for the new invoice rooms. The old ones with 2 CREATED report actions still need to be cleaned up.

**Comment by truph01** [comment]
@cristipaval The bug "Now I understand the issue. There are 2 CREATED report actions for the invoice report PLUS the one that belongs to the transaction thread. I could reproduce." is not fully fixed. Step to reproduce:
1. In invoice room, create a new invoice to receiver.

2. Click on report preview to open the invoice report. Observe that there are 2 CREAD report actions in here.

3. Go to Troubleshoot > Clear cache and restart. 

4. Open the invoice report again. Observer that there is 1 CREATED report actions in here.



![image](https://github.com/user-attachments/assets/b213e95a-9018-40c2-8dca-95a568248785)

**Comment by truph01** [comment]
@rayane-djouah I push the latest changes in the draft PR.

Also, I updated [comment](https://github.com/Expensify/App/issues/43797#issuecomment-2301585972) to include my suggestion changes.

**Comment by truph01** [comment]
@rayane-djouah Ah, my bad. Beside ```createdIOUReportActionID``` and ```createdReportActionIDForThread```, I need to send ```reportActionID``` to BE as well. Now it works well.

@cristipaval Plz ignore my [comment](https://github.com/Expensify/App/issues/43797#issuecomment-2301585972) above.

**Comment by cristipaval** [comment]
Waiting for the PR to get deployed.

**Comment by dylanexpensify** [comment]
Bump @cristipaval 🙇‍♂️

**Comment by cristipaval** [comment]
> @cristipaval - Reminder that the old reports with 2 CREATED report actions still need to be cleaned up. Are you tracking this internally?

I don't think it makes sense to spend time on cleaning that as we don't have active users using the invoice rooms yet.

@madmax330, do you think we should run some custom queries to clean the DB, or should we deal with it if it becomes a real problem?

**Comment by madmax330** [comment]
> do you think we should run some custom queries to clean the DB, or should we deal with it if it becomes a real problem?

@cristipaval are they real users or test users?
Let's clean them up anyway so we don't spend time looking into it later?

**Comment by dylanexpensify** [comment]
@madmax330 shall we get a new engineer on this? Or are you on it?

**Comment by madmax330** [comment]
@dylanexpensify actually I think we can just close this out since it was fixed already and the 2 report actions don't cause much of a problem

**Comment by madmax330** [comment]
Cool is there a payment summary already?

**Comment by dylanexpensify** [comment]
Payment summary:

Contributor: @truph01 $250 
Contributor+: @rayane-djouah $250 

Please apply/request!

**Comment by dylanexpensify** [comment]
on it now.

**Comment by dylanexpensify** [comment]
done!

**Comment by dylanexpensify** [comment]
Ah sorry @truph01, paid!
